crates/devices/virtio-vsock/src/packet.rs

```diff
@@ -0,0 +1,934 @@
+        self
+    }
+
+    /// Return the `op` of the header.
+    pub fn op(&self) -> u16 {
+        self.header_inner.op.into()
+    }
+
+    /// Set the `op` of the header.
+    pub fn set_op(&mut self, op: u16) -> &mut Self {
+        set_header_field!(self, op, OP_OFFSET, op);
+        self
+    }
+
+    /// Return the `flags` of the header.
+    pub fn flags(&self) -> u32 {
+        self.header_inner.flags.into()
+    }
+
+    /// Set the `flags` of the header.
+    pub fn set_flags(&mut self, flags: u32) -> &mut Self {
+        set_header_field!(self, flags, FLAGS_OFFSET, flags);
+        self
+    }
+
+    /// Set a specific flag of the header.
+    pub fn set_flag(&mut self, flag: u32) -> &mut Self {
+        self.set_flags(self.flags() | flag);
+        self
+    }
+
+    /// Return the `buf_alloc` of the header.
+    pub fn buf_alloc(&self) -> u32 {
+        self.header_inner.buf_alloc.into()
+    }
+
+    /// Set the `buf_alloc` of the header.
+    pub fn set_buf_alloc(&mut self, buf_alloc: u32) -> &mut Self {
+        set_header_field!(self, buf_alloc, BUF_ALLOC_OFFSET, buf_alloc);
+        self
+    }
+
+    /// Return the `fwd_cnt` of the header.
+    pub fn fwd_cnt(&self) -> u32 {
+        self.header_inner.fwd_cnt.into()
+    }
+
+    /// Set the `fwd_cnt` of the header.
+    pub fn set_fwd_cnt(&mut self, fwd_cnt: u32) -> &mut Self {
+        set_header_field!(self, fwd_cnt, FWD_CNT_OFFSET, fwd_cnt);
+        self
+    }
+
+    /// Create the packet wrapper from a TX chain.
+    ///
+    /// The chain head is expected to hold a valid packet header. A following packet data
+    /// descriptor can optionally end the chain.
+    pub fn from_tx_virtq_chain<M, T>(
+        mem: &'a M,
+        desc_chain: &mut DescriptorChain<T>,
+    ) -> Result<Self>
+    where
+        M: GuestMemory,
+        <<M as GuestMemory>::R as GuestMemoryRegion>::B: WithBitmapSlice<'a, S = B>,
+        T: Deref,
+        T::Target: GuestMemory,
+    {
+        let chain_head = desc_chain.next().ok_or(Error::DescriptorChainTooShort)?;
+        // All TX buffers must be device-readable.
+        if chain_head.is_write_only() {
+            return Err(Error::UnexpectedWriteOnlyDescriptor);
+        }
+
+        // The packet header should fit inside the buffer corresponding to the head descriptor.
+        if (chain_head.len() as usize) < PKT_HEADER_SIZE {
+            return Err(Error::DescriptorLengthTooSmall);
+        }
+
+        let header = mem
+            .get_slice(chain_head.addr(), PKT_HEADER_SIZE)
+            .map_err(Error::InvalidMemoryAccess)?;
+
+        let header_inner = mem
+            .read_obj(chain_head.addr())
+            .map_err(Error::InvalidMemoryAccess)?;
+
+        let mut pkt = Self {
+            header,
+            header_inner,
+            data: None,
+        };
+
+        // If the `len` field of the header is zero, then the packet doesn't have a `data` element.
+        if pkt.len() == 0 {
```

Comment:
I don't have a very strong opinion about this, but maybe it's worth it to implement `.is_empty()` and get rid of `#[allow(clippy::len_without_is_empty)]` if we are doing the empty check here.

Reply:
Good idea, I'll add it.